### DISCUSSION TITLE
Add additional default headers to $fetch ?

### DISCUSSION THREAD

**Discussion by guilledll** [nuxt/nuxt#17304] [Questions] [answered]
Hi community I have a problem with $fetch headers.

I have a **Laravel Backend** on **127.0.0.1:8000**, my **Nuxt 3** is hosted on **127.0.0.1:3000**. I'm using Laravel **Sanctum** for authentication and I'm having trouble because it uses cookies and an **x-xsrf-token** header to validate requests.

**THE NORMAL FLOW IS:**
- User make a request to 127.0.0.1:8000/sanctum/csrf-cookie
- In Nuxt I must get the cookies returned, set the in default headers for every next request and also set a new header called `x-xsrf-token` with the value of one of the cookies.
- Then user is able to do more requests...

In my nuxt app i'm able to see the response cookies sent by Laravel with a plugin I made:

```
import { splitCookiesString, parse } from "set-cookie-parser"

export default defineNuxtPlugin(async (nuxt) => {
	const res = await $fetch.raw("/sanctum/csrf-cookie", {
		baseURL: "http://127.0.0.1:8000"
	})

	for (const cookie of parse(splitCookiesString(res.headers.get("set-cookie")))) {
		useCookie(cookie.name).value = cookie.value
	}
})
```

With this then when i do `login`, the cookies are sent, **BUT** a header still missing from my request and **I don't know how to add it**.

My login is done in `index.vue`

```
<script setup>
const res = await useFetch("/login", {
	baseURL: "http://127.0.0.1:8000",
	method: "post",
	body: {
		email: "example@mail.com",
		password: "12345678"
	},
	headers: useRequestHeaders(["cookie"])
})
</script>
```
Using `useRequestHeaders` i'm able to attach cookies but NO way to attach the `x-xsrf-token` needed.

The token **MUST** be attached in the header and it will be excellent if a had a way to attach it to the **GLOBAL** instance of $fetch so i don't need to send it on every request.

Hope anyone can help, many of us use Laravel as server.

Thanks in advance!

**Comment by danielroe** [comment]
Would recommend you create a custom wrapper for `useFetch`, say in `~/composables/fetch.ts`:

```ts
export const useLaravelFetch = (url: string) => {
  return useFetch(url, {
	baseURL: "http://127.0.0.1:8000",
	method: "post",
	body: {
		email: "example@mail.com",
		password: "12345678"
	},
	headers: {
		'x-xsrf-token': 'your xsrf token',
		...useRequestHeaders(["cookie"])
    }
  })
}
```

**Reply by guilledll** [reply]
Thanks for the response.

The spread in the headers was useful but i still can't find the way to get the desired headers using fetch.

I'm still trying to solve this, my main issue now is that before doing the `/login` I need to do `/sanctum/csrf-cookie`, the thing is the headers returned from that endpoint are only accessible from server side because of the limitations of fetch, I get that. *(i searched why everywhere)*

How can I do a flow that the user press the **login button**, first call `/cookies` endpoint, get that headers to use cookies, then call the `/login` endpoint with the headers attached.

I can't find a way to call `/cookies` and `/login` in a single move and **from client**.

is it possible to do that with a plugin? like executing the plugin only when the user press the **login button**, that plugin needs to return the token value.

My plugin **to try**
```js
export default defineNuxtPlugin(async (nuxt) => {
	const response = await $fetch.raw("/sanctum/csrf-cookie", {
		baseURL: "http://127.0.0.1:8000"
	})

	let token = ""

	if (process.server) {
		const cookies = parse(splitCookiesString(response.headers.get("set-cookie")))

		const token = cookies.find((ck) => {
			return ck.name === "XSRF-TOKEN"
		})

		useCookie("x-xsrf-token").value = token.value
	}

	return {
		provide: {
			cookie: () => token // <--- this return is correct ?
		}
	}
})
```

Can i call the plugin in front to execute its function like this?
```js
<script setup>
const { $cookie } = useNuxtApp()

async function login() {
	const token = $cookie() // <-- get cookies here? how?
	console.log(token)
	await apiFetch("/login", {
		....
	})
}
```
With that in my execution the `plugin is not being executed` before the login.

Thanks in advance.

**Comment by Diizzayy** [comment]
@Guilledll To expand upon @danielroe's answer

> I need to do /sanctum/csrf-cookie

Within the custom fetch composable, you can check if a csrf token is present or not and act accordingly ( make the request to fetch it )

> How can I do a flow that the user press the login button, first call /cookies endpoint, get that headers to use cookies, then call the /login endpoint with the headers attached.

The code that I've written below is capable of doing just that, calling `useFetchCsrf('/example')` would first check if the `csrf cookie` is present, and will attempt to fetch the cookie if it isn't before making the intended request to `/example`

I've commented the code the explain what it does

`/composables/fetchCsrf.ts`
```ts
export const useFetchCsrf = async (...[url, opts]: Parameters<typeof $fetch>) => {
  // csrf token name
  const csrf_cookie = "XSRF-TOKEN";

  // check for 'XSRF-TOKEN' cookie
  let csrf = useCookie(csrf_cookie);

  // if 'XSRF-TOKEN' cookie isn't set, fetch and attempt to apply it
  if (!csrf.value) {
    let token: string = null;

    // if request is succesful, 'XSRF-TOKEN' cookie should be set
    const csrfRes = await $fetch.raw("/sanctum/csrf-cookie");

    if (process.server) {
      // On server, extract the csrf from the response
      const cookies = csrfRes.headers.get("set-cookie");
      token = cookies.split("XSRF-TOKEN=")[1].split(";")[0];
    } else {
      // on client-side, the cookie should now be present in your browser
      // but nuxt isn't aware since the cookie was updated outside it's context
      // so we need to manually update our variable.
      token = useCookie(csrf_cookie)?.value;
    }

    if (token) {
      // In SSR mode, this also sets the cookie on the client-side,
      // allowing it to be used for subsequent requests.
      csrf.value = token;
    }
  }

  const headers: HeadersInit = {
    // if headers were provided, apply them
    ...(opts?.headers && opts.headers),

    // only needed if proxied cookies are required in SSR
    // cookie: useRequestHeaders(['cookie'])?.cookie,

    // if we have a token, set 'x-xsrf-token' header
    ...(csrf.value && { "x-xsrf-token": csrf.value }),
  };

  return $fetch(url, { method: "POST", headers });
}
```

Example use: 

`app.vue`
```vue
<script lang="ts" setup>
async function login() {
  const res = await useFetchCsrf("/login", {
    method: "POST",
    body: {
      email: "example@mail.com",
      password: "12345678",
    },
  });
}
</script>
```

Also, note that if you wish to use this custom fetch for SSR requests ( initial request upon page load ), you'll need to wrap the the function with `useAsyncData`

**Comment by guilledll** [comment, marked as answer]
## Laravel sanctum + Nuxt 3 solution

First I want to thank @Diizzayy  for such a detailed answer.

I was able to **solve my original problem** by combining and changing some things from @Diizzayy and @danielroe answers.

I opted for a **client side** solution that allows me to use **Laravel Sanctum** with **Nuxt 3**, details next:

### Csrf token call

First a `fetch.ts` composable to allow calling my **laravel api** from browser using the `useApi` composable
```js
import { FetchOptions } from "ohmyfetch";

const serverRoute: string = "http://127.0.0.1:8000";
const apiRoute: string = `${serverRoute}/api`;
const csrf_cookie: string = "XSRF-TOKEN";

/**
 * Return the cookies needed by "Sanctum", browser will handle them automatically.
 */
export const useFetchCookies = async () => {
	await $fetch.raw("/sanctum/csrf-cookie", {
		baseURL: serverRoute,
		credentials: "include" // Allow browser to handle cookies
	});
};

/**
 * Api call using nuxt `useFetch`
 *
 * @see {@link https://github.com/unjs/ohmyfetch#readme} ~ ohmyfetch Docs
 * @param url
 * @param options
 */
export const useApi = async (url: string, options?: FetchOptions) => {
	// First we verify if the `xsrf-token` is present on the browser cookies
	let token = useCookie(csrf_cookie)?.value;

	if (!token) {
		// If not present we will re fetch all cookies, the browser will
		// handle them automatically so we don't need to do anything
		await useFetchCookies();

		// Load the new token value to use it in the `headers`
		token = useCookie(csrf_cookie).value;
	}

	// Here we will create a default set of headers for every request
	// if present we will also spread the `headers` set by the user
	// then we will delete them to avoid collision in next spread
	const headers: HeadersInit = {
		Accept: "application/json",
		"Cache-Control": "no-cache",
		"X-XSRF-TOKEN": token,
		...options?.headers
	};

	// At this point all the `headers` passed by the user where correctly
	// set in the defaults, now we will spread `options` to remove the
	// `headers` attribute so we don't spread it again in `useFetch`
	const opts: FetchOptions = options ? (({ headers, ...opts }) => opts)(options) : null;

	const baseURL = !options?.baseURL ? apiRoute : options.baseURL;

	return useFetch(url, {
		baseURL: baseURL,
		credentials: "include", // Allow browser to handle cookies
		headers,
		...opts
	});
};
```
### Calling login endpoint

From a Vue component now i'm able to use `/login` and also get the cookies (if missing) in one move, just by pressing a button.

```js
<script setup>
async function login() {
	await useApi("/login", {
		method: "POST",
		baseURL: "http://127.0.0.1:8000",
		body: {
			email: "example@mail.com",
			password: "password"
		}
	});
}
</script>

<template>
	<div>
		<button @click="login">login</button>
	</div>
</template>
```

### Hope this help other Laravel + Nuxt 3 devs!!!

**Reply by tiagomatosweb** [reply]
Hi, thanks for sharing this.

Are you able to perform SSR ajax call with Nuxt 3 and sanctum using either useAsyncData or useFetch?

**Reply by guilledll** [reply]
@tiagomatosweb i didn't find the way, that is why i created this useApi function.

**Reply by tiagomatosweb** [reply]
Yeah, I am struggling with that too.  Client side works perfectly, but SSR not much.

**Reply by cstriuli** [reply]
I'm using this composable and everything works fine but cannot request the `/api/user` (to fetch the user data). Always got a "401 Unauthorized" response, it's like the Authorization header it's missing or something but the `/login` request works perfect, it creates the cookie but it doesn't append it to the subsequent requests. Any idea?

**Reply by danielroe** [reply]
Check out https://v3.nuxtjs.org/guide/features/data-fetching/#isomorphic-fetch-and-fetch.

**Reply by cstriuli** [reply]
Thanks @danielroe ! Yep, i just tried that but it just returns an empty object :(

**requests log**
<img width="915" alt="Screen Shot 2022-07-09 at 18 00 01" src="https://user-images.githubusercontent.com/3141550/178122502-b8b52325-4629-4c87-95b8-2688f094f0a1.png">

**useRequestHeaders log**
<img width="380" alt="Screen Shot 2022-07-09 at 18 06 40" src="https://user-images.githubusercontent.com/3141550/178122635-220074ab-f274-4015-a61b-a791ec0fa126.png">


**cookies**
<img width="681" alt="Screen Shot 2022-07-09 at 18 05 37" src="https://user-images.githubusercontent.com/3141550/178122608-9376c025-8d52-4220-beb6-bbc95fb9c1df.png">

**Reply by danielroe** [reply]
It will return an empty object on client-side. The composable is for use on server side.

**Reply by cstriuli** [reply]
Oh, but i'm using the composable from @Guilledll which is client-side only.

So the login workflow (client-side) should be:
1. fetch csrf token ✅
2. login user ✅
3. fetch data ❌ 401 - main issue

Then, after a user reloads the page, a plugin fetches the user data from the server-side to persist the session or smth like that.

**Reply by amrnn90** [reply]
Hi @cstriuli, 
You might wanna check this functional starter I created for using Nuxt3 with Laravel:
https://github.com/amrnn90/breeze-nuxt

Specifically the implementation in this file might help you:
https://github.com/amrnn90/breeze-nuxt/blob/master/utils/%24larafetch.ts

**Reply by cstriuli** [reply]
It works @amrnn90 ! Amazing 🤩 thanks 🙏
